File: dotnet/SlimLO/Internal/WorkerProcess.cs
using System.Diagnostics;
using System.Text;
using System.Text.Json;

namespace SlimLO.Internal;

/// <summary>
/// Manages the lifecycle of a single native slimlo_worker subprocess.
/// Handles startup, message exchange, stderr capture, and crash detection.
/// </summary>
internal sealed class WorkerProcess : IAsyncDisposable
{
    private readonly string _workerPath;
    private readonly string _resourcePath;
    private readonly IReadOnlyList<string>? _fontDirectories;
    private Process? _process;
    private readonly SemaphoreSlim _lock = new(1, 1);
    private readonly StringBuilder _stderrBuffer = new();
    private volatile bool _disposed;
    private volatile bool _initialized;
    private int _conversionCount;
    private string? _version;

    public WorkerProcess(string workerPath, string resourcePath, IReadOnlyList<string>? fontDirectories)
    {
        _workerPath = workerPath;
        _resourcePath = resourcePath;
        _fontDirectories = fontDirectories;
    }

    public int ConversionCount => _conversionCount;
    public bool IsAlive => _process is { HasExited: false };
    public string? Version => _version;

    /// <summary>
    /// Start the worker process and send the init message.
    /// </summary>
    public async Task StartAsync(CancellationToken ct)
    {
        ObjectDisposedException.ThrowIf(_disposed, this);

        var psi = new ProcessStartInfo
        {
            FileName = _workerPath,
            UseShellExecute = false,
            RedirectStandardInput = true,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            CreateNoWindow = true,
            StandardInputEncoding = null, // binary
            StandardOutputEncoding = null, // binary
        };

        // Ensure the worker can find libslimlo and libmergedlo.
        // Include both the worker's directory and the resource program/ directory,
        // since the worker may be in a different location than the native assets
        // (e.g., during development or when SLIMLO_WORKER_PATH is set).
        var workerDir = Path.GetDirectoryName(_workerPath)!;
        var programDir = Path.Combine(_resourcePath, "program");
        if (OperatingSystem.IsLinux())
        {
            var existing = Environment.GetEnvironmentVariable("LD_LIBRARY_PATH");
            var paths = workerDir == programDir ? workerDir : $"{workerDir}:{programDir}";
            psi.Environment["LD_LIBRARY_PATH"] = string.IsNullOrEmpty(existing)
                ? paths
                : $"{paths}:{existing}";
        }
        else if (OperatingSystem.IsMacOS())
        {
            var existing = Environment.GetEnvironmentVariable("DYLD_LIBRARY_PATH");
            var paths = workerDir == programDir ? workerDir : $"{workerDir}:{programDir}";
            psi.Environment["DYLD_LIBRARY_PATH"] = string.IsNullOrEmpty(existing)
                ? paths
                : $"{paths}:{existing}";
        }

        // Set environment variables for the worker process.
        // On macOS, the SVP (headless) plugin is not available — only the Quartz
        // backend exists. On Linux/Windows, use SVP for headless operation.
        if (!OperatingSystem.IsMacOS())
            psi.Environment["SAL_USE_VCLPLUGIN"] = "svp";

        // On macOS, LOKit must use "unipoll" mode to run VCL initialization on the
        // calling thread. Without this, LOKit spawns a background thread for VCL init,
        // but the Quartz backend creates NSWindow objects which require the main thread.
        if (OperatingSystem.IsMacOS())
            psi.Environment["SAL_LOK_OPTIONS"] = "unipoll";

        // Enable font-related logging so stderr capture can detect font warnings
        psi.Environment["SAL_LOG"] = "+WARN.vcl.fonts+INFO.vcl+WARN.vcl";

        // Set custom font paths
        if (_fontDirectories is { Count: > 0 })
        {
            var separator = OperatingSystem.IsWindows() ? ";" : ":";
            psi.Environment["SAL_FONTPATH"] = string.Join(separator, _fontDirectories);
        }

        // Create a temp directory for LOKit user profile
        var profileDir = Path.Combine(Path.GetTempPath(), $"slimlo_profile_{Environment.ProcessId}_{GetHashCode():x}");
        Directory.CreateDirectory(profileDir);
        psi.Environment["HOME"] = profileDir;

        _process = Process.Start(psi)
            ?? throw new SlimLOException("Failed to start worker process", SlimLOErrorCode.InitFailed);

        // Start capturing stderr asynchronously
        _process.ErrorDataReceived += OnStderrData;
        _process.BeginErrorReadLine();

        // Send init message
        var initRequest = new InitRequest
        {
            ResourcePath = _resourcePath,
            FontPaths = _fontDirectories
        };
        var initBytes = Protocol.Serialize(initRequest);
        await Protocol.WriteMessageAsync(
            _process.StandardInput.BaseStream, initBytes, ct).ConfigureAwait(false);

        // Read init response
        var responseBytes = await Protocol.ReadMessageAsync(
            _process.StandardOutput.BaseStream, ct).ConfigureAwait(false);

        if (responseBytes is null)
        {
            var exitCode = _process.HasExited ? _process.ExitCode : -1;
            var stderr = GetStderrOutput();
            var message = $"Worker process died during initialization (exit code: {exitCode}).";

            if (stderr.Contains("error while loading shared libraries"))
            {
                message += " Missing system library detected. On Ubuntu/Debian, install: " +
                    "apt-get install libfontconfig1 libfreetype6 libexpat1 libcairo2 libpng16-16 " +
                    "libjpeg-turbo8 libxml2 libxslt1.1 libicu74 libnss3 libnspr4. " +
                    "See https://github.com/nicobao/libreoffice-to-pdf#linux-system-dependencies for details.";
            }

            message += $" Stderr: {stderr}";
            throw new SlimLOException(message, SlimLOErrorCode.InitFailed);
        }

        using var doc = Protocol.Deserialize(responseBytes);
        var root = doc.RootElement;
        var type = root.GetProperty("type").GetString();

        if (type == "error")
        {
            var message = root.TryGetProperty("message", out var m) ? m.GetString() : "Unknown error";
            throw new SlimLOException(
                $"Worker initialization failed: {message}",
                SlimLOErrorCode.InitFailed);
        }

        if (type == "ready")
        {
            _version = root.TryGetProperty("version", out var v) ? v.GetString() : null;
            _initialized = true;
        }
        else
        {
            throw new SlimLOException(
                $"Unexpected init response type: {type}",
                SlimLOErrorCode.InitFailed);
        }
    }

    /// <summary>
    /// Execute a conversion. The caller must hold the pool semaphore.
    /// </summary>
    public async Task<ConversionResult> ConvertAsync(
        ConvertRequest request,
        TimeSpan timeout,
        CancellationToken ct)
    {
        ObjectDisposedException.ThrowIf(_disposed, this);

        if (!_initialized || _process is null or { HasExited: true })
            return ConversionResult.Fail(
                "Worker process is not running",
                SlimLOErrorCode.NotInitialized, null);

        await _lock.WaitAsync(ct).ConfigureAwait(false);
        try
        {
            // Clear stderr buffer before conversion
            lock (_stderrBuffer)
                _stderrBuffer.Clear();

            // Create a linked cancellation token with timeout
            using var timeoutCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
            timeoutCts.CancelAfter(timeout);
            var linkedCt = timeoutCts.Token;

            try
            {
                // Send convert request
                var requestBytes = Protocol.Serialize(request);
                await Protocol.WriteMessageAsync(
                    _process.StandardInput.BaseStream, requestBytes, linkedCt).ConfigureAwait(false);

                // Read response
                var responseBytes = await Protocol.ReadMessageAsync(
                    _process.StandardOutput.BaseStream, linkedCt).ConfigureAwait(false);

                if (responseBytes is null)
                {
                    // Worker died during conversion
                    var exitCode = _process.HasExited ? _process.ExitCode : -1;
                    _initialized = false;
                    return ConversionResult.Fail(
                        $"Worker process crashed during conversion (exit code: {exitCode}). " +
                        "This typically indicates a malformed or corrupted document.",
                        SlimLOErrorCode.Unknown, null);
                }

                // Parse response
                using var doc = Protocol.Deserialize(responseBytes);
                var root = doc.RootElement;

                var diagnostics = root.TryGetProperty("diagnostics", out var diagArray)
                    ? StderrDiagnosticParser.ParseFromJson(diagArray)
                    : Array.Empty<ConversionDiagnostic>();

                var success = root.TryGetProperty("success", out var s) && s.GetBoolean();

                if (success)
                {
                    Interlocked.Increment(ref _conversionCount);
                    return ConversionResult.Ok(diagnostics);
                }
                else
                {
                    var errorMessage = root.TryGetProperty("error_message", out var em)
                        ? em.GetString() ?? "Conversion failed"
                        : "Conversion failed";
                    var errorCode = root.TryGetProperty("error_code", out var ec) && ec.ValueKind == JsonValueKind.Number
                        ? (SlimLOErrorCode)ec.GetInt32()
                        : SlimLOErrorCode.Unknown;

                    Interlocked.Increment(ref _conversionCount);
                    return ConversionResult.Fail(errorMessage, errorCode, diagnostics);
                }
            }
            catch (OperationCanceledException) when (timeoutCts.IsCancellationRequested && !ct.IsCancellationRequested)
            {
                // Timeout — kill the worker
                KillProcess();
                _initialized = false;
                return ConversionResult.Fail(
                    $"Conversion timed out after {timeout.TotalSeconds:F0} seconds",
                    SlimLOErrorCode.Unknown, null);
            }
        }
        finally
        {
            _lock.Release();
        }
    }

    /// <summary>
    /// Execute a buffer conversion. Sends document bytes, receives PDF bytes.
    /// The caller must hold the pool semaphore.
    /// </summary>
    public async Task<ConversionResult<byte[]>> ConvertBufferAsync(
        ConvertBufferRequest request,
        ReadOnlyMemory<byte> documentData,
        TimeSpan timeout,
        CancellationToken ct)
    {
        ObjectDisposedException.ThrowIf(_disposed, this);

        if (!_initialized || _process is null or { HasExited: true })
            return ConversionResult<byte[]>.Fail(
                "Worker process is not running",
                SlimLOErrorCode.NotInitialized, null);

        await _lock.WaitAsync(ct).ConfigureAwait(false);
        try
        {
            lock (_stderrBuffer)
                _stderrBuffer.Clear();

            using var timeoutCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
            timeoutCts.CancelAfter(timeout);
            var linkedCt = timeoutCts.Token;

            try
            {
                var stdin = _process.StandardInput.BaseStream;
                var stdout = _process.StandardOutput.BaseStream;

                // Send JSON header frame
                var requestBytes = Protocol.Serialize(request);
                await Protocol.WriteMessageAsync(stdin, requestBytes, linkedCt).ConfigureAwait(false);

                // Send binary document frame
                await Protocol.WriteMessageAsync(stdin, documentData, linkedCt).ConfigureAwait(false);

                // Read JSON response frame
                var responseBytes = await Protocol.ReadMessageAsync(stdout, linkedCt).ConfigureAwait(false);

                if (responseBytes is null)
                {
                    var exitCode = _process.HasExited ? _process.ExitCode : -1;
                    _initialized = false;
                    return ConversionResult<byte[]>.Fail(
                        $"Worker process crashed during buffer conversion (exit code: {exitCode}). " +
                        "This typically indicates a malformed or corrupted document.",
                        SlimLOErrorCode.Unknown, null);
                }

                using var doc = Protocol.Deserialize(responseBytes);
                var root = doc.RootElement;

                var diagnostics = root.TryGetProperty("diagnostics", out var diagArray)
                    ? StderrDiagnosticParser.ParseFromJson(diagArray)
                    : Array.Empty<ConversionDiagnostic>();

                var success = root.TryGetProperty("success", out var s) && s.GetBoolean();

                if (success)
                {
                    // Read binary PDF frame
                    var pdfBytes = await Protocol.ReadMessageAsync(stdout, linkedCt).ConfigureAwait(false);
                    if (pdfBytes is null)
                    {
                        _initialized = false;
                        return ConversionResult<byte[]>.Fail(
                            "Worker process crashed while sending PDF data",
                            SlimLOErrorCode.Unknown, diagnostics);
                    }

                    Interlocked.Increment(ref _conversionCount);
                    return ConversionResult<byte[]>.Ok(pdfBytes, diagnostics);
                }
                else
                {
                    var errorMessage = root.TryGetProperty("error_message", out var em)
                        ? em.GetString() ?? "Buffer conversion failed"
                        : "Buffer conversion failed";
                    var errorCode = root.TryGetProperty("error_code", out var ec) && ec.ValueKind == JsonValueKind.Number
                        ? (SlimLOErrorCode)ec.GetInt32()
                        : SlimLOErrorCode.Unknown;

                    Interlocked.Increment(ref _conversionCount);
                    return ConversionResult<byte[]>.Fail(errorMessage, errorCode, diagnostics);
                }
            }
            catch (OperationCanceledException) when (timeoutCts.IsCancellationRequested && !ct.IsCancellationRequested)
            {
                KillProcess();
                _initialized = false;
                return ConversionResult<byte[]>.Fail(
                    $"Buffer conversion timed out after {timeout.TotalSeconds:F0} seconds",
                    SlimLOErrorCode.Unknown, null);
            }
        }
        finally
        {
            _lock.Release();
        }
    }

    private void OnStderrData(object sender, DataReceivedEventArgs e)
    {
        if (e.Data is null) return;
        lock (_stderrBuffer)
        {
            _stderrBuffer.AppendLine(e.Data);
        }
    }

    private string GetStderrOutput()
    {
        lock (_stderrBuffer)
            return _stderrBuffer.ToString();
    }

    private void KillProcess()
    {
        try
        {
            if (_process is { HasExited: false })
                _process.Kill(entireProcessTree: true);
        }
        catch
        {
            // Best effort — process may have already exited
        }
    }

    public async ValueTask DisposeAsync()
    {
        if (_disposed) return;
        _disposed = true;

        if (_process is not null && !_process.HasExited)
        {
            try
            {
                // Try graceful shutdown first
                var quitBytes = Protocol.Serialize(new QuitRequest());
                await Protocol.WriteMessageAsync(
                    _process.StandardInput.BaseStream, quitBytes, CancellationToken.None)
                    .ConfigureAwait(false);

                // Wait up to 5 seconds for graceful exit
                using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(5));
                try
                {
                    await _process.WaitForExitAsync(cts.Token).ConfigureAwait(false);
                }
                catch (OperationCanceledException)
                {
                    KillProcess();
                }
            }
            catch
            {
                KillProcess();
            }
        }

        _process?.Dispose();
        _lock.Dispose();
    }
}

File: java/slimlo/src/main/java/com/slimlo/internal/WorkerProcess.java
package com.slimlo.internal;

import com.google.gson.JsonObject;
import com.slimlo.*;

import java.io.*;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.locks.ReentrantLock;

/**
 * Manages the lifecycle of a single native slimlo_worker subprocess.
 * Handles startup, message exchange, stderr capture, and crash detection.
 */
public final class WorkerProcess implements Closeable {

    private final String workerPath;
    private final String resourcePath;
    private final List<String> fontDirectories;
    private final ExecutorService executor;

    private Process process;
    private OutputStream stdin;
    private InputStream stdout;
    private final ReentrantLock lock = new ReentrantLock();
    private final StringBuilder stderrBuffer = new StringBuilder();
    private volatile boolean disposed;
    private volatile boolean initialized;
    private final AtomicInteger conversionCount = new AtomicInteger(0);
    private String version;

    public WorkerProcess(
            String workerPath,
            String resourcePath,
            List<String> fontDirectories,
            ExecutorService executor) {
        this.workerPath = workerPath;
        this.resourcePath = resourcePath;
        this.fontDirectories = fontDirectories;
        this.executor = executor;
    }

    public int getConversionCount() {
        return conversionCount.get();
    }

    public boolean isAlive() {
        return process != null && process.isAlive();
    }

    public String getVersion() {
        return version;
    }

    /**
     * Start the worker process and send the init message.
     */
    public void start() throws IOException, SlimLOException {
        if (disposed) throw new IllegalStateException("WorkerProcess is disposed");

        ProcessBuilder pb = new ProcessBuilder(workerPath);
        pb.redirectErrorStream(false);

        Map<String, String> env = pb.environment();

        // Library paths
        String workerDir = new File(workerPath).getParent();
        String programDir = new File(resourcePath, "program").getAbsolutePath();
        String libPaths = workerDir.equals(programDir) ? workerDir : workerDir + File.pathSeparator + programDir;

        if (WorkerLocator.isLinux()) {
            String existing = System.getenv("LD_LIBRARY_PATH");
            env.put("LD_LIBRARY_PATH", existing != null && !existing.isEmpty()
                    ? libPaths + ":" + existing : libPaths);
        } else if (WorkerLocator.isMacOS()) {
            String existing = System.getenv("DYLD_LIBRARY_PATH");
            env.put("DYLD_LIBRARY_PATH", existing != null && !existing.isEmpty()
                    ? libPaths + ":" + existing : libPaths);
        }

        // VCL plugin: SVP for headless on Linux/Windows, Quartz on macOS
        if (!WorkerLocator.isMacOS()) {
            env.put("SAL_USE_VCLPLUGIN", "svp");
        }

        // macOS: unipoll mode required for thread safety
        if (WorkerLocator.isMacOS()) {
            env.put("SAL_LOK_OPTIONS", "unipoll");
        }

        // Font logging
        env.put("SAL_LOG", "+WARN.vcl.fonts+INFO.vcl+WARN.vcl");

        // Custom font paths
        if (fontDirectories != null && !fontDirectories.isEmpty()) {
            String separator = WorkerLocator.isWindows() ? ";" : ":";
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < fontDirectories.size(); i++) {
                if (i > 0) sb.append(separator);
                sb.append(fontDirectories.get(i));
            }
            env.put("SAL_FONTPATH", sb.toString());
        }

        // Temp user profile
        String profileDir = System.getProperty("java.io.tmpdir") + File.separator
                + "slimlo_profile_" + Thread.currentThread().getId() + "_" + System.identityHashCode(this);
        new File(profileDir).mkdirs();
        env.put("HOME", profileDir);

        process = pb.start();
        stdin = process.getOutputStream();
        stdout = process.getInputStream();

        // Start stderr gobbler daemon
        startStderrGobbler();

        // Send init message
        Map<String, Object> initRequest = new HashMap<String, Object>();
        initRequest.put("type", "init");
        initRequest.put("resource_path", resourcePath);
        if (fontDirectories != null && !fontDirectories.isEmpty()) {
            initRequest.put("font_paths", fontDirectories);
        }

        byte[] initBytes = Protocol.serialize(initRequest);
        Protocol.writeMessage(stdin, initBytes);

        // Read init response
        byte[] responseBytes = Protocol.readMessage(stdout);
        if (responseBytes == null) {
            int exitCode = -1;
            try {
                if (process.waitFor(1, TimeUnit.SECONDS)) {
                    exitCode = process.exitValue();
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            String stderr = getStderrOutput();
            String message = "Worker process died during initialization (exit code: " + exitCode + ").";
            if (stderr.contains("error while loading shared libraries")) {
                message += " Missing system library detected. Check that all required native libraries are installed.";
            }
            message += " Stderr: " + stderr;
            throw new SlimLOException(message, SlimLOErrorCode.INIT_FAILED);
        }

        JsonObject root = Protocol.deserialize(responseBytes);
        String type = root.has("type") ? root.get("type").getAsString() : "";

        if ("error".equals(type)) {
            String message = root.has("message") ? root.get("message").getAsString() : "Unknown error";
            throw new SlimLOException("Worker initialization failed: " + message, SlimLOErrorCode.INIT_FAILED);
        }

        if ("ready".equals(type)) {
            version = root.has("version") ? root.get("version").getAsString() : null;
            initialized = true;
        } else {
            throw new SlimLOException("Unexpected init response type: " + type, SlimLOErrorCode.INIT_FAILED);
        }
    }

    /**
     * Execute a file-path conversion.
     */
    public ConversionResult convert(
            Map<String, Object> request,
            long timeoutMillis) {
        if (disposed) {
            return ConversionResult.fail("Worker is disposed", SlimLOErrorCode.NOT_INITIALIZED, null);
        }
        if (!initialized || process == null || !process.isAlive()) {
            return ConversionResult.fail("Worker process is not running", SlimLOErrorCode.NOT_INITIALIZED, null);
        }

        lock.lock();
        try {
            clearStderrBuffer();

            Future<ConversionResult> future = executor.submit(() -> {
                byte[] requestBytes = Protocol.serialize(request);
                Protocol.writeMessage(stdin, requestBytes);

                byte[] responseBytes = Protocol.readMessage(stdout);
                if (responseBytes == null) {
                    initialized = false;
                    int exitCode = process.isAlive() ? -1 : process.exitValue();
                    return ConversionResult.fail(
                            "Worker process crashed during conversion (exit code: " + exitCode + ").",
                            SlimLOErrorCode.UNKNOWN, null);
                }

                return parseConvertResponse(responseBytes, false);
            });

            try {
                return future.get(timeoutMillis, TimeUnit.MILLISECONDS);
            } catch (TimeoutException e) {
                future.cancel(true);
                killProcess();
                initialized = false;
                return ConversionResult.fail(
                        "Conversion timed out after " + (timeoutMillis / 1000) + " seconds",
                        SlimLOErrorCode.UNKNOWN, null);
            } catch (ExecutionException e) {
                Throwable cause = e.getCause();
                if (cause instanceof IOException) {
                    initialized = false;
                    return ConversionResult.fail(
                            "I/O error during conversion: " + cause.getMessage(),
                            SlimLOErrorCode.UNKNOWN, null);
                }
                return ConversionResult.fail(
                        "Conversion error: " + cause.getMessage(),
                        SlimLOErrorCode.UNKNOWN, null);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                return ConversionResult.fail("Conversion interrupted", SlimLOErrorCode.UNKNOWN, null);
            }
        } finally {
            lock.unlock();
        }
    }

    /**
     * Execute a buffer conversion. Sends document bytes, receives PDF bytes.
     */
    public ConversionResult convertBuffer(
            Map<String, Object> request,
            byte[] documentData,
            long timeoutMillis) {
        if (disposed) {
            return ConversionResult.fail("Worker is disposed", SlimLOErrorCode.NOT_INITIALIZED, null);
        }
        if (!initialized || process == null || !process.isAlive()) {
            return ConversionResult.fail("Worker process is not running", SlimLOErrorCode.NOT_INITIALIZED, null);
        }

        lock.lock();
        try {
            clearStderrBuffer();

            Future<ConversionResult> future = executor.submit(() -> {
                // Send JSON header frame
                byte[] requestBytes = Protocol.serialize(request);
                Protocol.writeMessage(stdin, requestBytes);

                // Send binary document frame
                Protocol.writeMessage(stdin, documentData);

                // Read JSON response frame
                byte[] responseBytes = Protocol.readMessage(stdout);
                if (responseBytes == null) {
                    initialized = false;
                    int exitCode = process.isAlive() ? -1 : process.exitValue();
                    return ConversionResult.fail(
                            "Worker process crashed during buffer conversion (exit code: " + exitCode + ").",
                            SlimLOErrorCode.UNKNOWN, null);
                }

                return parseConvertResponse(responseBytes, true);
            });

            try {
                return future.get(timeoutMillis, TimeUnit.MILLISECONDS);
            } catch (TimeoutException e) {
                future.cancel(true);
                killProcess();
                initialized = false;
                return ConversionResult.fail(
                        "Buffer conversion timed out after " + (timeoutMillis / 1000) + " seconds",
                        SlimLOErrorCode.UNKNOWN, null);
            } catch (ExecutionException e) {
                Throwable cause = e.getCause();
                if (cause instanceof IOException) {
                    initialized = false;
                }
                return ConversionResult.fail(
                        "Buffer conversion error: " + (cause != null ? cause.getMessage() : "unknown"),
                        SlimLOErrorCode.UNKNOWN, null);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                return ConversionResult.fail("Buffer conversion interrupted", SlimLOErrorCode.UNKNOWN, null);
            }
        } finally {
            lock.unlock();
        }
    }

    private ConversionResult parseConvertResponse(byte[] responseBytes, boolean isBuffer) throws IOException {
        JsonObject root = Protocol.deserialize(responseBytes);

        List<ConversionDiagnostic> diagnostics = root.has("diagnostics") && root.get("diagnostics").isJsonArray()
                ? StderrDiagnosticParser.parseFromJson(root.getAsJsonArray("diagnostics"))
                : Collections.<ConversionDiagnostic>emptyList();

        boolean success = root.has("success") && root.get("success").getAsBoolean();

        if (success) {
            conversionCount.incrementAndGet();
            if (isBuffer) {
                // Read binary PDF frame
                byte[] pdfBytes = Protocol.readMessage(stdout);
                if (pdfBytes == null) {
                    initialized = false;
                    return ConversionResult.fail(
                            "Worker process crashed while sending PDF data",
                            SlimLOErrorCode.UNKNOWN, diagnostics);
                }
                return ConversionResult.ok(pdfBytes, diagnostics);
            }
            return ConversionResult.ok(diagnostics);
        } else {
            String errorMessage = root.has("error_message") && !root.get("error_message").isJsonNull()
                    ? root.get("error_message").getAsString()
                    : "Conversion failed";
            SlimLOErrorCode errorCode = root.has("error_code") && root.get("error_code").isJsonPrimitive()
                    ? SlimLOErrorCode.fromValue(root.get("error_code").getAsInt())
                    : SlimLOErrorCode.UNKNOWN;

            conversionCount.incrementAndGet();
            return ConversionResult.fail(errorMessage, errorCode, diagnostics);
        }
    }

    private void startStderrGobbler() {
        final InputStream stderr = process.getErrorStream();
        Thread gobbler = new Thread(new Runnable() {
            @Override
            public void run() {
                try (BufferedReader reader = new BufferedReader(new InputStreamReader(stderr))) {
                    String line;
                    while ((line = reader.readLine()) != null) {
                        synchronized (stderrBuffer) {
                            stderrBuffer.append(line).append('\n');
                        }
                    }
                } catch (IOException e) {
                    // Process ended
                }
            }
        }, "slimlo-stderr-gobbler");
        gobbler.setDaemon(true);
        gobbler.start();
    }

    private void clearStderrBuffer() {
        synchronized (stderrBuffer) {
            stderrBuffer.setLength(0);
        }
    }

    private String getStderrOutput() {
        synchronized (stderrBuffer) {
            return stderrBuffer.toString();
        }
    }

    private void killProcess() {
        try {
            if (process != null && process.isAlive()) {
                process.destroyForcibly();
            }
        } catch (Exception e) {
            // Best effort
        }
    }

    @Override
    public void close() {
        if (disposed) return;
        disposed = true;

        if (process != null && process.isAlive()) {
            try {
                // Try graceful shutdown
                Map<String, Object> quitRequest = new HashMap<String, Object>();
                quitRequest.put("type", "quit");
                byte[] quitBytes = Protocol.serialize(quitRequest);
                Protocol.writeMessage(stdin, quitBytes);

                // Wait up to 5 seconds
                if (!process.waitFor(5, TimeUnit.SECONDS)) {
                    killProcess();
                }
            } catch (Exception e) {
                killProcess();
            }
        }

        if (process != null) {
            process.destroy();
        }
    }
}
